llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp

```diff
@@ -2868,3 +2889,883 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+    return false;
+  }
+  ValueBits *ICmpOp0Bits = Result->second;
+
+  // Now match the following cases
+  // (LSB): icmp [ne/eq] %mcrc, [1/0], where mcrc has LSB masked out
+  // (MSB): icmp [ne/eq] %mcrc, [1 << BitSize], where mcrc has MSB masked out
+  // (MSB): icmp [sgt/sge] %crc, [1/0]
+  // (MSB): icmp [slt/sle] %crc, [0/-1]
+  // And decide whether the check is checking for existence of 1 or 0
+  bool checkZero = false;
+  ValueBits::ValueBit *CheckBit = nullptr;
+  switch (Pred) {
+  case CmpInst::ICMP_NE:
+  case CmpInst::ICMP_EQ: {
+    // Check RHS is checking only one bit.
+    uint64_t RHSNum = RHS->getZExtValue();
+    uint64_t MSBNum = 1 << (ICmpOp0Bits->getSize() - 1);
+    // LSB if BitReversed, MSB otherwise.
+    if (!(CRC.BitReversed && RHSNum == 1) &&
+        !(!CRC.BitReversed && RHSNum == MSBNum) && RHSNum != 0) {
+      LLVM_DEBUG(dbgs() << DEBUG_TYPE
+                        << " CRCRegonize: ICmp RHS is not checking [M/L]SB\n");
+      return false;
+    }
+    // Now to check if we already know all the other bits of the RHS are zero.
+    ValueBits AllZeroValueBits((uint64_t)0, ICmpOp0Bits->getSize());
+    ValueBits *CRCOutBitsMasked = nullptr;
+    if (CRC.BitReversed) {
+      // Masking out the LSB is equivalent to shifting right one if we're just
+      // comparing all the other bits are zero.
+      CRCOutBitsMasked = ValueBits::LShr(ICmpOp0Bits, 1);
+      CheckBit = ICmpOp0Bits->getBit(0);
+    } else {
+      // The CRC type might be larger than the data, so we can't shift left
+      // one. Mask instead.
+      uint64_t MSBMask = ~(1 << (CRC.Width - 1));
+      CRCOutBitsMasked = ValueBits::And(ICmpOp0Bits, MSBMask);
+      CheckBit = ICmpOp0Bits->getBit(CRC.Width - 1);
+    }
+    if (!CRCOutBitsMasked->equals(&AllZeroValueBits)) {
+      LLVM_DEBUG(
+          dbgs() << DEBUG_TYPE
+                 << " CRCRegonize: Cannot determine ICmp checks [M/L]SB\n");
+      return false;
+    }
+    checkZero = RHSNum == 0;
+    break;
+  }
+  case CmpInst::ICMP_SGT:
+  case CmpInst::ICMP_SGE:
+  case CmpInst::ICMP_ULT:
+  case CmpInst::ICMP_ULE:
+    checkZero = true;
+    [[fallthrough]];
+  case CmpInst::ICMP_SLT:
+  case CmpInst::ICMP_SLE: {
+    int64_t RHSNum = RHS->getSExtValue();
+    if (((Pred == CmpInst::ICMP_SLT || Pred == CmpInst::ICMP_SGE) &&
+         RHSNum != 0) ||
+        ((Pred == CmpInst::ICMP_SLE) && RHSNum != -1) ||
+        ((Pred == CmpInst::ICMP_SGT) && RHSNum != 1) ||
+        ((Pred == CmpInst::ICMP_ULT) && RHSNum != (1 << (CRC.Width - 1))) ||
+        ((Pred == CmpInst::ICMP_ULE) && RHSNum != (1 << (CRC.Width - 1)) - 1)) {
+      LLVM_DEBUG(dbgs() << DEBUG_TYPE
+                        << " CRCRegonize: ICmp RHS is not checking MSB\n");
+      return false;
+    }
+    CheckBit = ICmpOp0Bits->getBit(CRCSize - 1);
+    break;
+  }
+  default:
+    return false;
+  }
+
+  // If there exists a Data input, ensure the check bit is crc^data.
```

Comment:
this comment is too specific to coremark.  Can you make it more general?

Reply:
Looking at many CRC implementations, I disagree that this is specific to CoreMark. The data is almost always xor'd into the crc. Half of these, coremark included, decide to xor inside the loop.